File: cli/src/main/java/nisui/cli/print_formats/PrintCSV.java
package nisui.cli.print_formats;

import java.io.PrintStream;

public class PrintCSV<V> extends PrintFormat<V> {
    @Override
    public void printHeader(PrintStream out) {
        boolean isFirst = true;
        for (Field field : fields) {
            if (isFirst) {
                isFirst = false;
            } else {
                out.print(',');
            }
            out.print(field.caption);
        }
        out.println();
    }

    @Override
    public void printValue(PrintStream out, V value) {
        boolean isFirst = true;
        for (Field field : fields) {
            if (isFirst) {
                isFirst = false;
            } else {
                out.print(',');
            }
            out.print(field.extract.apply(value));
        }
        out.println();
    }

    @Override
    public void printFooter(PrintStream out) {
    }
}

File: cli/src/main/kotlin/nisui/cli/print_formats/PrintCSV.kt
package nisui.cli.print_formats

import java.io.PrintStream

public class PrintCSV<V> : PrintFormat<V>() {
    override fun printHeader(out_: PrintStream) {
        var isFirst = true
        for (field in fields) {
            if (isFirst) {
                isFirst = false
            } else {
                out_.print(',')
            }
            out_.print(field.caption)
        }
        out_.println()
    }

    override fun printValue(out_: PrintStream, value: V) {
        var isFirst = true
        for (field in fields) {
            if (isFirst) {
                isFirst = false
            } else {
                out_.print(',')
            }
            out_.print(field.extract(value))
        }
        out_.println()
    }

    override fun printFooter(out_: PrintStream) {
    }
}
